File: docs/algorithms/code/algorithms/tree/recursion-traversal/inorder-traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function inorderTraversalHelper(root: TreeNode | null, arr: number[]) {
  if(!root) {
    return
  }

  if(root.left) {
    inorderTraversalHelper(root.left, arr);
  }

  arr.push(root.val);

  if(root.right) {
    inorderTraversalHelper(root.right, arr);
  }
}

function inorderTraversal(root: TreeNode | null): number[] {
  if(!root) {
    return [];
  }

  const arr: number[] = [];
  inorderTraversalHelper(root, arr);
  return arr;
};

File: docs/algorithms/code/algorithms/inorderTraversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func inorderTraversalHelper(root *TreeNode, arr *[]int) {
	if root == nil {
		return;
	}
	if root.Left!= nil {
		inorderTraversalHelper(root.Left, arr);
	}
	*arr = append(*arr, root.Val);
	if root.Right!= nil {
		inorderTraversalHelper(root.Right, arr);
	}
}

func inorderTraversal(root *TreeNode) []int {
	arr := make([]int, 0)

	if root == nil {
		return arr
	}
	
	inorderTraversalHelper(root, &arr)
	return arr
}